board.gd

```diff
@@ -0,0 +1,147 @@
+class_name MainBoard
+extends TileMap
+
+# Local variables.
+
+# Game dimensions.
+var dimensions = Vector2i(10, 20)
+
+# 2D game board to store piece positions and state.
+var board = []
+
+
+## Called when the node enters the scene tree for the first time.
+func _ready():
+	reset_board()
+
+
+func reset_board():
+	self.clear()
+	board.clear()
+	for x in range(dimensions.x):
+		board.append(Array())
+		for y in range(dimensions.y):
+			board[x].append(0)
+
+
+func Board_drawGhost(piece: Piece):
+	# clear map
+	clear_layer(1)
+	# move ghost to bottom until not clear or past bottom
+	var ghost_blocks = piece.blocks.map(func(block): return block + piece.position)
+	var found_row = false
+	while not found_row:
+		for block in ghost_blocks:
+			if block.y >= dimensions.y:
+				found_row = true
+			elif board[block.x][block.y]:
+				found_row = true
+		if not found_row:
+			for i in ghost_blocks.size():
+				ghost_blocks[i] += Vector2.DOWN
+		else:
+			for i in ghost_blocks.size():
+				ghost_blocks[i] += Vector2.UP
+
+	for block in ghost_blocks:
+		set_cell(1, Vector2i(block.x, block.y), 7, Vector2i(0, 0), 0)
```

Comment:
For future cleanup, we should probably use enums to represent the tile IDs. Something like:

```suggestion
		set_cell(1, Vector2i(block.x, block.y), GHOST_TILE, Vector2i(0, 0), 0)
```